File: localdb/src/main/kotlin/ru/tutu/stations/localdb/base/DbTransaction.java
package ru.tutu.stations.localdb.base;

import android.support.annotation.NonNull;

import java.util.concurrent.Callable;

/**
 * Транзакция БД
 *
 * @author Grigoriy Pryamov
 */
public interface DbTransaction {
    /**
     * Начинает транзакцию
     */
    void beginTransaction();

    /**
     * Завершает транзакцию
     */
    void endTransaction();

    /**
     * Помечает успешную транзакцию
     */
    void setTransactionSuccessful();

    /**
     * Выполняет блок кода в транзакции
     */
    <T> T callInTx(@NonNull Callable<T> callable);

    /**
     * Выполняет блок кода в транзакции
     */
    void callInTx(@NonNull Runnable runnable);
}

File: localdb/src/main/kotlin/ru/tutu/stations/localdb/base/DbTransaction.kt
package ru.tutu.stations.localdb.base

import java.util.concurrent.Callable

/**
 * @author Grigoriy Pryamov.
 */
interface DbTransaction {

    /**
     * Начинает транзакцию
     */
    fun beginTransaction()

    /**
     * Завершает транзакцию
     */
    fun endTransaction()

    /**
     * Помечает успешную транзакцию
     */
    fun setTransactionSuccessful()

    /**
     * Выполняет блок кода в транзакции
     */
    fun <T> callInTx(callable: Callable<T>): T

    /**
     * Выполняет блок кода в транзакции
     */
    fun callInTx(runnable: Runnable)
}